daq/host.py

```diff
@@ -559,37 +558,52 @@ def _docker_test(self, test_name):
-            'target_mac': self.target_mac,
-            'gateway_ip': self.gateway.host.IP(),
-            'gateway_mac': self.gateway.host.MAC(),
-            'inst_base': self._inst_config_path(),
-            'port_base': self._port_base,
-            'device_base': self._device_aux_path(),
-            'type_base': self._type_aux_path(),
-            'scan_base': self.scan_base
-        }
-        self.test_host = docker_test.DockerTest(self.runner, self.target_port, self.devdir,
-                                                test_name)
-        self.test_port = self.runner.allocate_test_port(self.target_port)
-        if self.ext_loip:
-            params['local_ip'] = self.ext_loip.replace('@', '%d') % self.test_port
-            params['switch_ip'] = self.switch_setup.get('ip_addr')
-            params['switch_port'] = str(self.target_port)
-            params['switch_model'] = self.switch_setup.get('model')
-            params['switch_username'] = self.switch_setup.get('username')
-            params['switch_password'] = self.switch_setup.get('password')
-
+        self.test_host = docker_test.DockerTest(self.runner, self.target_port,
+                                                self.devdir, test_name)
         try:
+            self.test_port = self.runner.allocate_test_port(self.target_port)
+            switch_setup = self.switch_setup if 'mods_addr' in self.switch_setup else None
+            ext_loip = switch_setup.get('mods_addr') % self.test_port if switch_setup else None
+
+            params = {
+                'local_ip': ext_loip,
+                'target_ip': self.target_ip,
+                'target_mac': self.target_mac,
+                'target_port': str(self.target_port),
+                'gateway_ip': self.gateway.host.IP(),
+                'gateway_mac': self.gateway.host.MAC(),
+                'inst_base': self._inst_config_path(),
+                'port_base': self._port_base,
+                'device_base': self._device_aux_path(),
+                'type_base': self._type_aux_path(),
+                'scan_base': self.scan_base
+            }
+            if ext_loip:
+                params.update(self._get_switch_config())
+
             LOGGER.debug('test_host start %s/%s', test_name, self._host_name())
-            self._set_module_config(self._loaded_config)
+            self._write_module_config(self._loaded_config, self._host_tmp_path())
+            self._record_result(self.test_name, config=self._loaded_config, state=MODE.CONF)
             self.record_result(test_name, state=MODE.EXEC)
             self._monitor_scan(os.path.join(self.scan_base, 'test_%s.pcap' % test_name))
             self.test_host.start(self.test_port, params, self._docker_callback, self._finish_hook)
-        except:
-            self.test_host = None
-            raise
+        except Exception as e:
+            if self.test_host:
+                self.test_host.terminate(optional=True)
```

Comment:
Is the optional necessary here? From what I can recall, the self.test_host is a direct indicator of whether  the docker_host has been terminated.

Reply:
I think you are half right. test_host is not a direct indicator because the docker_host doesn't exist until the test_host has been started, so during initialization there's a state where the test_host exists but docker_host does not; however, that doesn't mean "optional" was the best fix.  I've refactored the code to break it up into smaller functions, take out the optional, and catch errors at appropriate spots so it can more directly cleanup state.